hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/HazelcastSqlValidator.java

```diff
@@ -352,22 +352,39 @@ protected void validateOrderList(SqlSelect select) {
     protected void validateJoin(SqlJoin join, SqlValidatorScope scope) {
         super.validateJoin(join, scope);
 
+        boolean leftInputIsStream = containsStreamingSource(join.getLeft());
+        boolean rightInputIsStream = containsStreamingSource(join.getRight());
+
+        if (leftInputIsStream && rightInputIsStream) {
+            throw newValidationError(join, RESOURCE.streamToStreamJoinNotSupported());
+        }
+
         switch (join.getJoinType()) {
-            case INNER:
-            case COMMA:
-            case CROSS:
             case LEFT:
-                if (containsStreamingSource(join.getRight())) {
-                    throw newValidationError(join, RESOURCE.streamingSourceOnWrongSide());
+                if (rightInputIsStream) {
+                    throw newValidationError(join, RESOURCE.streamingSourceOnWrongSide("right", "LEFT"));
                 }
                 break;
             case RIGHT:
-                if (containsStreamingSource(join.getLeft())) {
-                    throw newValidationError(join, RESOURCE.streamingSourceOnWrongSide());
+                if (leftInputIsStream) {
+                    throw newValidationError(join, RESOURCE.streamingSourceOnWrongSide("left", "RIGHT"));
                 }
                 break;
             case FULL:
                 throw QueryException.error(SqlErrorCode.PARSING, "FULL join not supported");
+            case INNER:
+                // Batch-to-stream INNER join, swap sides to convert it to stream-to-batch join, which is supported.
```

Comment:
The swapping from batch-to-stream to stream-to-batch isn't implemented. And if it was, then INNER, COMMA and CROSS would work too, COMMA and CROSS is just a different syntax for INNER join.

Reply:
https://github.com/hazelcast/hazelcast/issues/20486